juju/juju
# CAAS operator sets agent version on startup, restarts when api addresses change

## Description of change

Three key functional additions:
1. caas operator sets agent version metadata in application state on startup
(similar to what unit agents do for their units)
2. controller api address changes (eg going to HA) trigger an operator restart
(and hence new addresses are used)
3. juju upgrades will trigger caas operator restart with an updated docker image

Also a driveby fix for permission checks removing a caas unit.

A fair bit of boilerplate in this PR to add the necessary facade API changes. Not a great deal of logic changes. The tools metadata setting uses a common API plugin. The API address watching also uses a common API plugin.

## QA steps

bootstrap caas model and deploy a charm
observe tools metadata being set
enable ha and observe operator restart
exec into the operator and check the agent file for new addresses
upgrade juju and observe an operator restart using new docker image
